ocaml-lsp-server/src/code_actions/action_inline.ml

```diff
@@ -0,0 +1,119 @@
+open Import
+
+let action_title = "Inline"
+
+type inline_task =
+  { ident : Ident.t
+  ; body : Typedtree.expression  (** the expression to inline *)
+  ; context : Typedtree.expression  (** where to perform inlining *)
+  }
+
+let find_inline_task pipeline pos =
+  let contains loc pos =
+    match Position.compare_inclusion pos (Range.of_loc loc) with
+    | `Outside _ -> false
+    | `Inside -> true
+  in
+
+  (* Find most enclosing nonrecursive let binding *)
+  let browse =
+    Mpipeline.typer_result pipeline
+    |> Mtyper.get_typedtree |> Mbrowse.of_typedtree
+  in
+
+  Mbrowse.enclosing
+    (Mpipeline.get_lexing_pos pipeline (Position.logical pos))
+    [ browse ]
+  |> List.find_map ~f:(function
+       | ( _
+         , Browse_raw.Expression
+             { exp_desc =
+                 Texp_let
+                   ( Nonrecursive
+                   , [ { vb_pat = { pat_desc = Tpat_var (id, s); _ }
+                       ; vb_expr
+                       ; _
+                       }
+                     ]
+                   , rhs )
+             ; _
+             } )
+         when contains s.loc pos ->
+         Some { ident = id; body = vb_expr; context = rhs }
+       | _ -> None)
+
+let find_parsetree_loc pipeline loc k =
+  let expr_iter (iter : Ocaml_parsing.Ast_iterator.iterator)
+      (expr : Parsetree.expression) =
+    if expr.pexp_loc = loc then k expr
+    else Ocaml_parsing.Ast_iterator.default_iterator.expr iter expr
+  in
+  let iterator =
+    { Ocaml_parsing.Ast_iterator.default_iterator with expr = expr_iter }
+  in
+  match Mpipeline.reader_parsetree pipeline with
+  | `Implementation s -> iterator.structure iterator s
+  | _ -> ()
+
+let inlined_text pipeline task =
+  let ret = ref None in
+  find_parsetree_loc pipeline task.body.exp_loc (fun expr ->
+      ret :=
+        Some (Format.asprintf "(%a)" Ocaml_parsing.Pprintast.expression expr));
+  Option.value_exn !ret
+
+(** Iterator over the text edits performed by the inlining task. *)
+let iter_inline_edits pipeline task k =
```

Comment:
I usually try to hide imperative code (use of iterator here) behind a functional interface -- we could have this function to simply return a list of edits and have a mutable state within itself hidden to its clients.